corehq/apps/api/resources/v0_5.py

```diff
@@ -1044,3 +1056,161 @@ def prepend_urls(self):
                 self._meta.resource_name), self.wrap_view('dispatch_list')),
         ]
+
+
+class NavigationEventAuditResource(HqBaseResource, ModelResource):
+    local_date = fields.DateField(attribute='local_date', readonly=True)
+    UTC_start_time = fields.DateTimeField(attribute='UTC_start_time', readonly=True)
+    UTC_end_time = fields.DateTimeField(attribute='UTC_end_time', readonly=True)
+
+    class Meta:
+        authentication = RequirePermissionAuthentication(HqPermissions.view_web_users)
+        queryset = NavigationEventAudit.objects.all()
+        resource_name = 'action_times'
+        fields = ['user']
+        include_resource_uri = False
+        allowed_methods = ['get']
+        detail_allowed_methods = []
+        limit = 10000
+        max_limit = 10000
+
+    # Compound filters take the form `prefix.qualifier=value`
+    # These filter functions are called with qualifier and value
+    COMPOUND_FILTERS = {
+        'local_date': make_date_filter(functools.partial(django_date_filter, field_name='local_date'))
+    }
+
+    @staticmethod
+    def to_obj(action_times):
+        '''
+        Takes a flat dict and returns an object
+        '''
+        return namedtuple('action_times', list(action_times))(**action_times)
+
+    def dispatch(self, request_type, request, **kwargs):
+        #super needs to be called first to authenticate user. Otherwise request.user returns AnonymousUser
+        response = super(HqBaseResource, self).dispatch(request_type, request, **kwargs)
+        if not toggles.ACTION_TIMES_API.enabled_for_request(request):
+            msg = (_("You don't have permission to access this API"))
+            raise ImmediateHttpResponse(HttpResponse(
+                json.dumps({"error": msg}),
+                content_type="application/json",
+                status=403))
+        else:
+            return response
+
+    def alter_list_data_to_serialize(self, request, data):
+        data['meta']['local_date_timezone'] = self.local_timezone.zone
+        params = request.GET.copy()  # Makes params mutable for creating next_url below
+
+        if data['meta']['total_count'] >= data['meta']['limit']:
+            last_object = data['objects'][-1]
+            params['cursor_local_date'] = last_object.data['local_date']
+            params['cursor_user'] = last_object.data['user']
+
+            next_url = '{}?{}'.format(request.path, params.urlencode())
+            data['meta']['next'] = next_url
+        return data
+
+    def dehydrate(self, bundle):
+        bundle.data['user_id'] = get_user_id_by_username(bundle.data['user'])
+        return bundle
+
+    def obj_get_list(self, bundle, **kwargs):
+        domain = kwargs['domain']
+        params = self._process_params(domain, bundle.request.GET)
+
+        results = self.cursor_query(domain, self.local_timezone, params)
+        return list(map(self.to_obj, results))
+
+    def _process_params(self, domain, params):
```

Comment:
Do you mean something like this?
```
@dataclass
class APIParams:
    users: List[str]
    limit: int
    local_timezone: str

    def validate(self):
        if self.limit < 0:
            raise ValueError("Limit cannot be negative.")

def process_params(domain, params):
    param_values = {
        'users': params.getlist('users'),
        'limit': int(params.get('limit', 0)),
        'local_timezone': params.get('local_timezone'),
    }

    api_params = APIParams(**param_values)
    api_params.validate()
    
    return {
        'users': api_params.users,
        'limit': api_params.limit,
        'local_timezone': local_timezone,
    }
```

Reply:
Yeah, something like that, and I'd even then pass around the `api_params` object rather than a `dict`, so the other code can work with a dataclass object rather than a dict - usually just a little nicer to work with.  Totally just a suggestion though - take it or leave it.

I'd also put in something to catch unexpected params just to head off confusion if someone makes a request like `?user=myusername@example.com` (ie, in my opinion, that should fail with an error about `user` not being a valid parameter, rather than ignoring that filter).

Reply:
I was working on incorporating this but was running into a couple problems. Can we move forward with this if everything else looks fine, and i'll spin up another PR to refactor this